fix(vertex/anthropic): passing beta header only for structured outputs

aayush-kapoor opened the pull request:
## Background

Currently, the vertex anthropic provider only checked the model ID and would pass the beta header for all models that allow structured outputs.

But vertex doesn't yet support passing these beta headers. So we would get an [error like this](https://github.com/vercel/ai/issues/10981) even for a simple generateText function

This PR also addresses a bug when strict mode is enabled and the header is passed (shouldn't happen)

## Summary

- only pass the beta header when actually using native output format and when using strict tools

## Manual Verification

the example `google-vertex-anthropic-sonnet-4-5.ts` was used to verify the fix (was failing before)

## Checklist

- [x] Tests have been added / updated (for bug fixes / features)
- [ ] Documentation has been added / updated (for bug fixes / features)
- [x] A _patch_ changeset for relevant packages has been added (for bug fixes / features - run `pnpm changeset` in the project root)

## Future Work

ideally, we handle this shortcoming in our implementation of vertex provider
- if beta header is to be passed, we intercept it and then we could pass a providerOption saying `structuredOutputMode: 'jsonTool'` (continued in this PR after the comment https://github.com/vercel/ai/pull/11015#issuecomment-3634356904)

## Related Issues

Fixes #10981

aayush-kapoor committed:
failing tests

aayush-kapoor committed:
failing example

aayush-kapoor committed:
fix

aayush-kapoor committed:
example changed

aayush-kapoor commented:
all the commits till here fix the original issue. 

commits after this point are to account for the change in example `/examples/ai-core/src/generate-object/google-vertex-anthropic.ts` which fails in the current flow

that example will fail and give an error. but to enable it for users who want to use the latest anthropic models with vertex, we force the use of JSON tool

aayush-kapoor committed:
fix for structured outputs in vertex

aayush-kapoor committed:
cs

aayush-kapoor commented on examples/ai-core/src/generate-object/google-vertex-anthropic-sonnet-4-5.ts:
was failing before fix

aayush-kapoor committed:
Merge branch 'main' into aayush/10981-vertex-structured-outputs

lgrammel commented on packages/anthropic/src/anthropic-messages-language-model.test.ts:
```diff
@@ -2335,7 +2529,7 @@ describe('AnthropicMessagesLanguageModel', () => {
           expect(server.calls[0].requestHeaders).toMatchInlineSnapshot(`
             {
-              "anthropic-beta": "structured-outputs-2025-11-13,advanced-tool-use-2025-11-20",
+              "anthropic-beta": "advanced-tool-use-2025-11-20",
```
why is this dropped?

lgrammel reviewed (commented):
code looks good but test changes indicate a regression might have been introduced

aayush-kapoor committed:
Merge branch 'main' into aayush/10981-vertex-structured-outputs

aayush-kapoor commented on packages/anthropic/src/anthropic-messages-language-model.test.ts:
```diff
@@ -2441,7 +2635,7 @@ describe('AnthropicMessagesLanguageModel', () => {
           expect(server.calls[0].requestHeaders).toMatchInlineSnapshot(`
             {
-              "anthropic-beta": "structured-outputs-2025-11-13,advanced-tool-use-2025-11-20",
+              "anthropic-beta": "advanced-tool-use-2025-11-20",
```
same as above, expected behaviour

aayush-kapoor commented:
> code looks good but test changes indicate a regression might have been introduced

expected behaviour, see comment above!

aayush-kapoor committed:
rm file

aayush-kapoor committed:
Merge branch 'main' into aayush/10981-vertex-structured-outputs

aayush-kapoor committed:
Merge branch 'main' into aayush/10981-vertex-structured-outputs

aayush-kapoor committed:
test regression fix

gr2m commented on .changeset/lemon-sheep-behave.md:
why two changeset files?

aayush-kapoor replied:
will merge into one - created while iterating

gr2m reviewed (approved):
code changes look good but haven't tested end-to-end, I assume you did

aayush-kapoor committed:
Merge branch 'main' into aayush/10981-vertex-structured-outputs

aayush-kapoor committed:
edit cs

Ensure beta header is passed only for structured outputs in vertex and anthropic.

aayush-kapoor committed:
Delete .changeset/lemon-sheep-behave.md

aayush-kapoor commented:
> code changes look good but haven't tested end-to-end, I assume you did

not with UI example (there isn't one for vertex or structured outputs) but yes, tested with different variations of examples (passing tools, tools + object output)

aayush-kapoor committed:
Merge branch 'main' into aayush/10981-vertex-structured-outputs

aayush-kapoor commented:
> This also seems to fix a bug around strict mode and setting the header. Can this either be separated out into a different PR or added to the description?

adding in the description
